# Исследование надежности заемщиков


In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

Выведем количество пропущенных значений для каждого столбца.

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Вывела строчки с пустыми значениями в столбцах days_employed и total_income. Их общее количество 2174, что равно количеству пропусков в каждом по отдельности столбцо. Значит, в строках с пропускам значения отстутствуют в обоих этих столбцах.

In [ ]:
data.loc[(data['days_employed'].isna()) & (data['total_income'].isna()) ]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`.

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

Встречается отрицательное количество дней трудового стажа в столбце `days_employed`. Обработаем значения в этом столбце с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях.

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

Выведем перечень уникальных значений столбца `children`.

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма `data`.

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

Заполним пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Убедимся, что все пропуски заполнены.

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

Заменим вещественный тип данных в столбце `total_income` на целочисленный.

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру. Проверем остальные столбцы.

In [ ]:
data['education'] = data['education'].str.lower()

Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследование данных

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Чтобы посмотреть существует ли зависимость возврата кредита от количества детей, необходимо посмотреть процентное соотношение возврата кредита для каждого количества детей.

В столбце debt 0 - отсутствие задолженности у клиента, 1 - ее наличие. Таким образом, мы можем посчитать количество людей с долгами по каждой категории количества детей.

In [ ]:


data.groupby('children')['debt'].sum() / data.groupby('children')['debt'].count() * 100

children
0    7.543822
1    9.234609
2    9.454191
3    8.181818
4    9.756098
5    0.000000
Name: debt, dtype: float64

Из процентного соотношения видно, что должников с одним ребенком, двумя и четырьмя детьмя примерно одинаковое количество и их большинство среди остальных. Чуть лучше кредиты возвращют те, у кого 3 детей, и еще лучше кредиторы без детей. Вероятно, это связано с отсутствием трат на ребенка.

То, что среди людей с 5 детьми 0% должников по кредиту выглядит странно. Необходимо проверить достаточность выборки.

In [ ]:
display(data[data['children'] == 5])

children  days_employed  dob_years  education  education_id  \
3979          5    1572.328285         42    среднее             1   
4397          5    3248.839837         36    среднее             1   
7866          5     773.124856         36    среднее             1   
15822         5     418.199982         31    среднее             1   
15916         5    2286.262752         37    среднее             1   
16211         5     387.317579         35    среднее             1   
20452         5     268.425464         38  начальное             3   
20837         5    2386.600221         35    среднее             1   
21156         5    1690.018117         59    среднее             1   

          family_status  family_status_id gender  income_type  debt  \
3979   гражданский брак                 1      M    сотрудник     0   
4397    женат / замужем                 0      F    компаньон     0   
7866    женат / замужем                 0      F    сотрудник     0   
15822   женат / замужем                 0      F    сотрудник     0   
15916   женат / замужем                 0      F    сотрудник     0   
16211  гражданский брак                 1      F  госслужащий     0   
20452   женат / замужем                 0      F    сотрудник     0   
20837   женат / замужем                 0      F    компаньон     0   
21156   женат / замужем                 0      M    сотрудник     0   

       total_income                           purpose total_income_category  \
3979         142594      на покупку своего автомобиля                     C   
4397         168460          операции с недвижимостью                     C   
7866          48772                 операции с жильем                     D   
15822         77552  сделка с подержанным автомобилем                     C   
15916        256698              покупка недвижимости                     B   
16211        126102             на проведение свадьбы                     C   
20452        212545      заняться высшим образованием                     B   
20837        204241                             жилье                     B   
21156        269068   операции со своей недвижимостью                     B   

               purpose_category  
3979     операции с автомобилем  
4397   операции с недвижимостью  
7866   операции с недвижимостью  
15822    операции с автомобилем  
15916  операции с недвижимостью  
16211        проведение свадьбы  
20452     получение образования  
20837  операции с недвижимостью  
21156  операции с недвижимостью

Видно, что в базе только 9 записей о людях с 5 детьми и их возвратом кредита. В данном случае сложно оценить зависимость возврата кредита от количества детей.

**Вывод:** 

Таким образом, существует определенная зависимость между количеством детей и возвратом кредита в срок. Люди без детей, берущие кредит, чаще не имеют задолженностей, чем люди, у которых есть дети. 

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Здесь и в дальнейшем воспользуемся алгоритмом из предыдущего задания.

In [ ]:
data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['debt'].count() * 100

family_status
Не женат / не замужем    9.763948
в разводе                7.064760
вдовец / вдова           6.624606
гражданский брак         9.313014
женат / замужем          7.560558
Name: debt, dtype: float64

Исходя из полученных соотношений, можно сделать вывод, что люди не состоящие в браке имеют тенденцию к задержкам выплат. Вероятно, это связано с тем, что их работа - их единственный источник дохода. При этом люди, состоящие в гражданском браке также имеют близкий процент задержек к "одиночкам". Это может быть связано с устройством их совместного бюджета.
Люди, состоящие в браке, и люди в разводе имеют приблизительно одинаковый процент задолженностей по кредиту.
Меньше всего задолженностей у вдовцов и вдов.

**Вывод:** 

Исходя из результатов можно сказать, что существует определенная зависимость между семейным положением и возвратом кредита в срок. Меньше всего задержек по возврату возникает у вдовцов и вдов.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
data.groupby('total_income_category')['debt'].sum() / data.groupby('total_income_category')['debt'].count() * 100

total_income_category
A    8.000000
B    7.060231
C    8.498210
D    6.017192
E    9.090909
Name: debt, dtype: float64

Из полученных соотношений видно, что больше всего задолженностей по возврату кредита у людей с доходом категории Е (0 - 30 000). Немноо меньше задолженностей у категории С (50 000 - 200 000) и категории A (больше 1 000 000). Самый маленький роцент задолженностей у категории D (50 001 - 200 000).

**Вывод:** 

Таким образом, существует отпределенная зависимость между уровнем дохода и возвратом кредита в срок. Лучше всего кредит возвращают в срок люди с доходом от 50 000 до 200 000 рублей.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['debt'].count() * 100

purpose_category
операции с автомобилем      9.347978
операции с недвижимостью    7.255139
получение образования       9.252758
проведение свадьбы          7.911803
Name: debt, dtype: float64

Из полученных соотношений видно, что кредиты для операций с автомобилями и получения образования имеют больше задолженностей. Самый маленький процент задолженностей в категории операции с недвижемостью.

**Вывод:** 

Таким образом, существует определенная зависимость между целью кредита и его возвратом в срок. Лучше всего кредит возвращают в срок, если он предназначен для операций с недвижимостью.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

Т.к. значения отсутствуют одновременно в двух столбцах, вероятно имел место технический сбой.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* медианные значение характеризуют именно середину из всего возможного диапазона количественных переменных, и они не будут сбиваться экремумами.

### Шаг 4: общий вывод.

Было проведено исследование влияния различных факторов на факт погашения кредита в срок. Исходные данные представлены в виде статистики платежеспособности и других важных для исследования факторов из личных характеристик клиента. Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Перед проведением исследования были выдвинуты несколько гипотез:
* *Гипотеза 1*. Между количеством детей и возвратом кредита в срок есть зависимость
* *Гипотеза 2*. Между семейным положением и возвратом кредита в срок есть зависимость
* *Гипотеза 3*. Между уровнем дохода и возвратом кредита в срок есть зависимость
* *Гипотеза 4*. Между целью кредита и его возвратом в срок есть зависимость

В результате удалось выявить зависимости.

**Гипотеза 1 частично подтверждена**: * для более точных выводов необходимо больше информации по многодетным семьям (3 и более детей).

* 7,5% - доля просроченных кредитов у заемщиков без детей;
* 9,2% - доля просроченных кредитов у заемщиков с 1 ребенком;
* 9,4% -доля просроченных кредитов у заемщиков с 2 детьми;

Рекомендация: провести дополнительное исследование на большей выборке данных.

**Гипотеза 2 подтверждена:**

* У неженатых клиентов (9,8%) и тех, у кого официально не зарегистрирован брак (9,3%), выше вероятность стать должниками.
* Те, кто в разводе (7,1%), женат/замужем (7,5%) или вдовец/вдова (6,6%) менбше вероятность стать задолжниками.

Рекомендация: Разниуа между показателями не слишком существенна, поэтому не рекоммендуется придавать днной зависимости большой вес при выношении решения о выдаче кредита.

**Гипотеза 3 частично подтеверждена:**

* 7,1% - доля просроченных кредитов у заемщиков с уровнем дохода от 200 000 до 1 000 000;
* 8,5% - доля просроченных кредитов у заемщиков с уровнем дохода от 50 000 до 200 000.

**Гипотеза 4 подтверждена:**

* 7,3%- доля просроченных кредитов, целью которых были операции с недвижимостью;
* 7,9% - доля просроченных кредитов, целью которых было проведение свадьбы;
* 9,3% - доля просроченных кредитов, целью которых были операции с автомобилями;
* 9,3% - доля просроченных кредитов, целью которых было получение образования.

Рекомендация: разница между пунктами 1 и 4 списка всего 2%, поэтому не рекоммендутся придавать днной зависимости большой вес при выношении решения о выдаче кредита.

**Общие рекоммендации:**

1. При построении скрининговой системы можно использовать зависимости из второй и четвертой гипотезы (семейное положение и цели кредита).
2. Необходимы дополнительные данные для повторной выборки для проверки второй и третьей гипотезы.